### ISSUE TITLE
qubes services crashes (SEGV) in debian-9 template

### ISSUE THREAD

**Issue by marmarek** [QubesOS/qubes-issues#4349]
### Qubes OS version:

R4.0

### Affected component(s): 

qubesdb
qrexec

---

### Steps to reproduce the behavior:

On affected hardware: start debian-9 template or a qube based on it.

### Expected behavior:

Normal startup, able to start applications.

### Actual behavior:

```
systemd[1]: Starting Qubes DB agent...
systemd[1]: Started Apply Kernel Variables.
systemd[1]: Mounted Configuration File System.
kernel: qubesdb-daemon[275]: segfault at 775bd4a2eff8 ip 0000775bd481bf64 sp 0000775bd4a2f000 error 6 in ld-2.24.so[775bd4812000+23000]
systemd[1]: qubes-db.service: Main process exited, code=killed, status=11/SEGV
systemd[1]: Failed to start Qubes DB agent.
systemd[1]: qubes-db.service: Unit entered failed state.
systemd[1]: qubes-db.service: Failed with result 'signal'.
```

```
kernel: qrexec-agent[568]: segfault at 72851f15fff8 ip 000072851ef5c355 sp 000072851f160000 error 6 in ld-2.24.so[72851ef46000+23000]
```

Only debian-9 is affected, fedora works fine.

**Comment by marmarek** [comment]
Inspecting coredump, it looks like a stack overflow (hitting guard page) in xenstore read thread (internally created by libxenstore). The actual crash is during runtime symbol loading while the thread is canceled (`xs_close` calls `pthread_cancel` on that thread). Full backtrace:
```
#0  0x00007ffff7de2f64 in _dl_lookup_symbol_x (undef_name=0x7ffff51b4c63 "_Unwind_Find_FDE", undef_map=0x55555575b8c0, ref=ref@entry=0x7ffff7ff1028, 
    symbol_scope=0x55555575bc18, version=0x55555575be10, type_class=type_class@entry=1, flags=5, skip_map=0x0) at dl-lookup.c:810
#1  0x00007ffff7de7c54 in _dl_fixup (l=<optimized out>, reloc_arg=<optimized out>) at ../elf/dl-runtime.c:111
#2  0x00007ffff7def35a in _dl_runtime_resolve_xsavec () at ../sysdeps/x86_64/dl-trampoline.h:125
#3  0x00007ffff51c1b13 in uw_frame_state_for (context=context@entry=0x7ffff7ff1ca0, fs=fs@entry=0x7ffff7ff1af0)
    at ../../../src/libgcc/unwind-dw2.c:1241
#4  0x00007ffff51c2d30 in uw_init_context_1 (context=context@entry=0x7ffff7ff1ca0, outer_cfa=outer_cfa@entry=0x7ffff7ff1ed0, 
    outer_ra=0x7ffff6ab0d60 <__GI___pthread_unwind+64>) at ../../../src/libgcc/unwind-dw2.c:1562
#5  0x00007ffff51c3336 in _Unwind_ForcedUnwind (exc=0x7ffff7ff3d70, stop=stop@entry=0x7ffff6ab0bd0 <unwind_stop>, stop_argument=0x7ffff7ff2c50)
    at ../../../src/libgcc/unwind.inc:201
#6  0x00007ffff6ab0d60 in __GI___pthread_unwind (buf=<optimized out>) at unwind.c:121
#7  0x00007ffff6aa6c5a in __do_cancel () at ./pthreadP.h:283
#8  sigcancel_handler (sig=<optimized out>, si=0x7ffff7ff2030, ctx=<optimized out>) at nptl-init.c:220
#9  <signal handler called>
#10 0x00007ffff79116ed in read () at ../sysdeps/unix/syscall-template.S:84
#11 0x00007ffff5ddc867 in read (__nbytes=16, __buf=0x7ffff0000a30, __fd=6) at /usr/include/x86_64-linux-gnu/bits/unistd.h:44
#12 read_all (fd=6, data=data@entry=0x7ffff0000a30, len=len@entry=16, nonblocking=0) at xs.c:377
#13 0x00007ffff5ddc94f in read_all (len=16, nonblocking=0, data=0x7ffff0000a30, fd=<optimized out>) at xs.c:1149
#14 read_message (h=h@entry=0x55555575b620, nonblocking=0) at xs.c:1149
#15 0x00007ffff5ddcc81 in read_thread (arg=0x55555575b620) at xs.c:1221
#16 0x00007ffff6aa8494 in start_thread (arg=0x7ffff7ff3700) at pthread_create.c:333
#17 0x00007ffff791eacf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
```

`_dl_runtime_resolve_xsavec` made me look at xsave, kernel prints: `x86/fpu: Enabled xstate features 0xff, context size is 2560 bytes, using 'compacted' format.` This is quite big, given it's 960 on my Skylake-based machine.

There are two workarounds:
1. Disable xsave (`noxsave` kernel opt).
2. Run affected processes with `LD_BIND_NOW=1` to avoid calling runtime loader in that thread.

**Comment by marmarek** [comment]
[This upstream commit](https://github.com/xen-project/xen/commit/448c03b3cbe14873ee637755a29ea26ee7ca9ef9) looks like fix for this very problem.

**Comment by brendanhoar** [comment]
> Inspecting coredump, it looks like a stack overflow (hitting guard page) in xenstore read thread (internally created by libxenstore). The actual crash is during runtime symbol loading while the thread is canceled (`xs_close` calls `pthread_cancel` on that thread). Full backtrace:
> 
> ```
> #0  0x00007ffff7de2f64 in _dl_lookup_symbol_x (undef_name=0x7ffff51b4c63 "_Unwind_Find_FDE", undef_map=0x55555575b8c0, ref=ref@entry=0x7ffff7ff1028, 
>     symbol_scope=0x55555575bc18, version=0x55555575be10, type_class=type_class@entry=1, flags=5, skip_map=0x0) at dl-lookup.c:810
> #1  0x00007ffff7de7c54 in _dl_fixup (l=<optimized out>, reloc_arg=<optimized out>) at ../elf/dl-runtime.c:111
> #2  0x00007ffff7def35a in _dl_runtime_resolve_xsavec () at ../sysdeps/x86_64/dl-trampoline.h:125
> #3  0x00007ffff51c1b13 in uw_frame_state_for (context=context@entry=0x7ffff7ff1ca0, fs=fs@entry=0x7ffff7ff1af0)
>     at ../../../src/libgcc/unwind-dw2.c:1241
> #4  0x00007ffff51c2d30 in uw_init_context_1 (context=context@entry=0x7ffff7ff1ca0, outer_cfa=outer_cfa@entry=0x7ffff7ff1ed0, 
>     outer_ra=0x7ffff6ab0d60 <__GI___pthread_unwind+64>) at ../../../src/libgcc/unwind-dw2.c:1562
> #5  0x00007ffff51c3336 in _Unwind_ForcedUnwind (exc=0x7ffff7ff3d70, stop=stop@entry=0x7ffff6ab0bd0 <unwind_stop>, stop_argument=0x7ffff7ff2c50)
>     at ../../../src/libgcc/unwind.inc:201
> #6  0x00007ffff6ab0d60 in __GI___pthread_unwind (buf=<optimized out>) at unwind.c:121
> #7  0x00007ffff6aa6c5a in __do_cancel () at ./pthreadP.h:283
> #8  sigcancel_handler (sig=<optimized out>, si=0x7ffff7ff2030, ctx=<optimized out>) at nptl-init.c:220
> #9  <signal handler called>
> #10 0x00007ffff79116ed in read () at ../sysdeps/unix/syscall-template.S:84
> #11 0x00007ffff5ddc867 in read (__nbytes=16, __buf=0x7ffff0000a30, __fd=6) at /usr/include/x86_64-linux-gnu/bits/unistd.h:44
> #12 read_all (fd=6, data=data@entry=0x7ffff0000a30, len=len@entry=16, nonblocking=0) at xs.c:377
> #13 0x00007ffff5ddc94f in read_all (len=16, nonblocking=0, data=0x7ffff0000a30, fd=<optimized out>) at xs.c:1149
> #14 read_message (h=h@entry=0x55555575b620, nonblocking=0) at xs.c:1149
> #15 0x00007ffff5ddcc81 in read_thread (arg=0x55555575b620) at xs.c:1221
> #16 0x00007ffff6aa8494 in start_thread (arg=0x7ffff7ff3700) at pthread_create.c:333
> #17 0x00007ffff791eacf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
> ```
> `_dl_runtime_resolve_xsavec` made me look at xsave, kernel prints: `x86/fpu: Enabled xstate features 0xff, context size is 2560 bytes, using 'compacted' format.` This is quite big, given it's 960 on my Skylake-based machine.
> 
> There are two workarounds:
> 
> 1. Disable xsave (`noxsave` kernel opt).
> 2. Run affected processes with `LD_BIND_NOW=1` to avoid calling runtime loader in that thread.

Clarification, if you don't mind?
1. Is the noxsave kernel option for dom0 or the debian VM?
2. Can an example be given applying `LD_BIND_NOW=1` in practice for a VM that boots but fails to register with dom0 and is killed after timeout?

What I'm seeing (I've included testing repo for dom0 and all templates): after updates earlier this week, Debian continued to work, but the whonix VMs did not (they boot fine according to the log, but dom0 kills them because it doesn't think they started correctly). If the issue effects whonix as well, perhaps I didn't update whonix and debian in parallel.

I will most likely reinstall the whonix templates via: manually unwire them from VMs/global settings; sudo dnf remove; sudo qubesctl salt method of download/installation). 

Brendan

**Comment by marmarek** [comment]
>     1. Is the noxsave kernel option for dom0 or the debian VM?

Debian VM, for example `qvm-prefs debian-9 kernelopts noxsave`
As for `LD_BIND_NOW=1`, it's not that trivial, because you need to put that in multiple places, so I'd go fist with `noxsave`.

**Comment by brendanhoar** [comment]
Thanks! I tried nopat noxsave on both but it did not resolve. I must be having a separate issue. I'll reinstall the whonix-14 templates.

**Comment by jrsmiley** [comment]
Just out of curiosity was this an X299 chipset?
